### PR TITLE
feat: add generator option proto-plus-dep

### LINKED ISSUE #1611: Incorrect assumption in gapic-generator-python for dependencies
By default, python clients (gapics) are generated with [proto-plus](https://github.com/googleapis/proto-plus-python) types.

There is an incorrect assumption in the generator where it assumes that dependencies never have proto-plus types.

For example, see the comment below which shows the problem. All dependencies are treated as non-proto plus (with `_pb2`) extension. When gapics depend on other gapics, the generator produces broken code.

https://github.com/googleapis/gapic-generator-python/blob/0ef9e62a64457bc9ca710524cad3993a58721e0b/gapic/schema/metadata.py#L92-L96

The code for `is_external_type` is below
https://github.com/googleapis/gapic-generator-python/blob/0ef9e62a64457bc9ca710524cad3993a58721e0b/gapic/schema/metadata.py#L105-L107

Googlers see b/266837434

### LINKED ISSUE #10828: cannot run maps-fleetengine-delivery-v1
in the beta libary pip install https://storage.googleapis.com/fleetengine-gapic/dist/793/2022-07-19/maps-fleetengine-delivery-v1-py.tar.gz I am getting error as below :+1: 

ModuleNotFoundError: No module named 'google.geo.type'

when in install google-geo-type 0.3.0 module then I get the next error as 

ImportError: cannot import name 'viewport_pb2' from 'google.geo.type

Reference :
[https://developers.google.com/maps/documentation/transportation-logistics/last-mile-fleet-solution/fleet-performance/fleet-engine/gapic_client](https://developers.google.com/maps/documentation/transportation-logistics/last-mile-fleet-solution/fleet-performance/fleet-engine/gapic_client)

### PR DESCRIPTION
**Pull request by parthea**
fix: resolve ImportError when GAPIC depends on another GAPIC

Fixes b/266837434
Fixes https://github.com/googleapis/google-cloud-python/issues/10828
Fixes https://github.com/googleapis/gapic-generator-python/issues/1611

Googlers see go/fix-gapic-depends-on-gapic-with-proto-plus-types

### CODE REVIEW

File: gapic/schema/metadata.py
Comment (vchudnov-g) [inline gapic/schema/metadata.py:200]:
```suggestion
            version_regex = "^v\d+(alpha|beta)?\d*"
```

Comment (vchudnov-g) [reply gapic/schema/metadata.py:200]: We could have double digit versions on either side of the preview labels.

Also, we have seen some more complicated version numbers like [these](https://github.com/googleapis/googleapis/tree/master/google/cloud/asset) (v1p5beta1). So maybe the regex should be `"^v\d[^/]*$"` ? WDYT?

File: gapic/templates/setup.py.j2
Comment (vchudnov-g) [inline gapic/templates/setup.py.j2:36]: The TODO comment above this `{%if}` should be updated, since https://github.com/googleapis/gapic-generator-python/pull/1240 has been merged.

File: gapic/templates/setup.py.j2
Comment (vchudnov-g) [inline gapic/templates/setup.py.j2:38]: Could you file an issue and refer to it in a TODO to have a more elegant way of handling these cases? It seems wrong to hard-code packages and versions in the template (brittle, technical debt).

File: gapic/templates/testing/constraints-3.7.txt.j2
Comment (vchudnov-g) [inline gapic/templates/testing/constraints-3.7.txt.j2:16]: Ditto for issue/TODO here.

File: gapic/schema/metadata.py
Comment (vchudnov-g) [inline gapic/schema/metadata.py:203]: Do we have a test for this import regex rewriting? We should, for at least cases like `v1`, `v10`,`v2beta`,`v20beta``,v20p1`,`v20p1alpha3p5`, don't you think?

**Review by vchudnov-g — CHANGES_REQUESTED** [review]
Don't know much about proto-plus so I can't comment on those aspects of it, but otherwise this looks good.

Two important issues: making the regex more robust, and capturing an item to reduce the hard-coding tech debt for specific API×version.

File: gapic/schema/metadata.py
Comment (parthea) [reply gapic/schema/metadata.py]: Fixed

File: gapic/templates/setup.py.j2
Comment (parthea) [reply gapic/templates/setup.py.j2]: Fixed

File: gapic/templates/setup.py.j2
Comment (parthea) [reply gapic/templates/setup.py.j2]: I've moved these cases to a separate file and extracted the common code.

File: gapic/templates/testing/constraints-3.7.txt.j2
Comment (parthea) [reply gapic/templates/testing/constraints-3.7.txt.j2]: Fixed

File: gapic/schema/metadata.py
Comment (parthea) [reply gapic/schema/metadata.py:203]: Done

File: gapic/templates/setup.py.j2
Comment (vchudnov-g) [inline gapic/templates/setup.py.j2:35]: Could you clarify `the package is different from this setup.py`? This to check we're not processing specifying a version range for the GAPIC we're generating itself, only for others we may depend on?

**Review by vchudnov-g — APPROVED** [review]
One test suggestion and one question, but otherwise LGTM

File: gapic/templates/setup.py.j2
Comment (parthea) [reply gapic/templates/setup.py.j2:35]: Correct. There is an existing test that shows the `requires_package` [method](https://github.com/googleapis/gapic-generator-python/blob/13b6c6c1da9b75c171cc53167cd048aba2dc9481/gapic/schema/api.py#L487) will return true for the API that is currently being generated.
https://github.com/googleapis/gapic-generator-python/blob/13b6c6c1da9b75c171cc53167cd048aba2dc9481/tests/unit/schema/test_api.py#L93-L111

**Comment by parthea** [comment]
Waiting on approval in go/fix-gapic-depends-on-gapic-with-proto-plus-types

File: gapic/schema/metadata.py
Comment (atulep) [inline gapic/schema/metadata.py:106]: Simplify by returning the condition directly.

return if ....

File: gapic/schema/metadata.py
Comment (atulep) [inline gapic/schema/metadata.py:105]: This is non-trivial function and will benefit from adding a docstring explaining what and why you are checking for these things. Please add.

File: gapic/schema/metadata.py
Comment (atulep) [inline gapic/schema/metadata.py:204]: Can a package have less than 2 elements? (so [-2] raise IndexError)

File: gapic/schema/metadata.py
Comment (parthea) [reply gapic/schema/metadata.py]: Fixed in 
https://github.com/googleapis/gapic-generator-python/pull/1596/commits/99a1ac8da58713d758b508f5980f7a940d228773

File: gapic/schema/metadata.py
Comment (parthea) [reply gapic/schema/metadata.py]: Done in https://github.com/googleapis/gapic-generator-python/pull/1596/commits/47a2067690019bde6e17e0025ced16325ec72692

File: gapic/schema/metadata.py
Comment (parthea) [reply gapic/schema/metadata.py:214]: Good catch! Fixed in https://github.com/googleapis/gapic-generator-python/pull/1596/commits/3f95c0af570cd68db0dc634f79a5760573d3f406

**Pull request merged**
